File: leetcode/24_SwapNodesInPairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pairExists = head != None and head.next != None
        if pairExists: # we have something to swap; otherwise skip swap
            first = head.next
            second = head
            last = head.next.next
            first.next = second
            second.next = swapPairs(self, last)
            return second

        return head # could be null or a single item

File: leetcode/24_SwapNodesInPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        boolean pairExists = head != null && head.next != null;
        if (pairExists) {
            // We have something to swap; otherwise skip swap
            ListNode first = head.next;
            ListNode second = head;
            ListNode last = head.next.next;
            first.next = second;
            second.next = swapPairs(last);
            return first;
        }
        return head; // could be null or a single item
    }
}